modin/pandas/test/dataframe/test_indexing.py

```diff
@@ -886,6 +887,132 @@ def test_reset_index(data):
     df_equals(modin_df_cp, pd_df_cp)
 
 
+@pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
+@pytest.mark.parametrize("nlevels", [3])
+@pytest.mark.parametrize("columns_multiindex", [True, False])
+@pytest.mark.parametrize(
+    "level",
+    ["no_level", None, 0, 1, 2, [2, 0], [2, 1], [1, 0], [2, 1, 2], [0, 0, 0, 0]],
+)
+@pytest.mark.parametrize("col_level", ["no_col_level", 0, 1, 2])
+@pytest.mark.parametrize(
+    "col_fill", ["no_col_fill", None, 0, "col_one", 222.222, "new"]
+)
+@pytest.mark.parametrize("drop", [False])
+@pytest.mark.parametrize("multiindex_levels_names_max_levels", [0, 1, 2, 3, 4])
+@pytest.mark.parametrize(
+    "none_in_index_names", [False, True, "mixed_1st_None", "mixed_2nd_None"]
+)
+def test_reset_index_with_multi_index_no_drop(
+    data,
+    nlevels,
+    columns_multiindex,
+    level,
+    col_level,
+    col_fill,
+    drop,
+    multiindex_levels_names_max_levels,
+    none_in_index_names,
+):
+    data_rows = len(data[list(data.keys())[0]])
+    index = generate_multiindex(data_rows, nlevels=nlevels)
+    data_columns = len(data.keys())
+    columns = (
+        generate_multiindex(data_columns, nlevels=nlevels)
+        if columns_multiindex
+        else pandas.RangeIndex(0, data_columns)
+    )
+    index.names = (
+        [f"level_{i}" for i in range(index.nlevels)]
+        if multiindex_levels_names_max_levels == 0
+        else [
+            tuple(
+                [
+                    f"level_{i}_name_{j}"
+                    for j in range(
+                        0,
+                        max(multiindex_levels_names_max_levels + 1 - index.nlevels, 0)
+                        + i,
+                    )
+                ]
+            )
+            if max(multiindex_levels_names_max_levels + 1 - index.nlevels, 0) + i > 0
+            else f"level_{i}"
+            for i in range(index.nlevels)
+        ]
+    )
+
+    if none_in_index_names is True:
+        index.names = [None] * len(index.names)
+    elif none_in_index_names:
+        names_list = list(index.names)
+        start_index = 0 if none_in_index_names == "mixed_1st_None" else 1
+        names_list[start_index::2] = [None] * len(names_list[start_index::2])
+        index.names = names_list
+
+    modin_df = pd.DataFrame(data, index=index, columns=columns)
+    pandas_df = pandas.DataFrame(data, index=index, columns=columns)
```

Comment:
Notice that when we're specifying custom columns at the DataFrame constructor it does not replace labels of the passed dict, but doing reindex:
```python
>>> data = {f"col{i}": np.arange(5) for i in range(4)}
>>> columns = ["a", "b", "col0", "c"]
>>> pandas.DataFrame(data, columns=columns)
     a    b  col0    c
0  NaN  NaN     0  NaN
1  NaN  NaN     1  NaN
2  NaN  NaN     2  NaN
3  NaN  NaN     3  NaN
4  NaN  NaN     4  NaN
>>> columns[2] = "d"
>>> pandas.DataFrame(data, columns=columns)
Empty DataFrame
Columns: [a, b, d, c]
Index: []
```
So, in terms of this test, it seems that we're always creating an empty frame.

We should either explicitly set index columns after DataFrame was created `df.columns = columns`, or convert data-dict to the 2D array to discard column labels.

Reply:
Thank you, I wasn't aware of this behavior of DataFrame constructor. I fixed test to modify `data` with generated column keys and it seems to work as intended now. A good thing that this behavior was debugged already on a "simple" dataframe which I later removed as redundant. So no new bugs were found after I fixed the test.